ufl/sorting.py

```diff
@@ -126,7 +127,7 @@ def cmp_expr(a, b):
 
         # Now we know that the type is the same, check further based
         # on type specific properties.
-        if a._ufl_is_terminal_:
+        if a._ufl_is_terminal_ and not isinstance(a, ExternalOperator):
```

Comment:
Again here. An `ExternalOperator` is both terminal and not terminal simultaneously.

Reply:
This does not say that ExternalOperator is not a terminal, it just says that it is a terminal that needs a special care. One of the consequence of having a terminal with operands is that we have to handle all the situations where thing were done on the assumption that Terminal do not have operands (so direct consequence of the choice done on the definition of ExternalOperator)